### ISSUE TITLE
Metered latency orders of magnitude too high?

### ISSUE THREAD

**Issue by steveblackburn** [dacapobench/dacapobench#283]
A user reported the following result.  I've not reproduced it, but I'd like to understand what's happening.

```
===== DaCapo simple tail latency: 50% 186 usec, 90% 264 usec, 99% 418 usec, 99.9% 1519 usec, 99.99% 3039 usec, max 9565 usec, measured over 1179648 events =====
===== DaCapo metered tail latency: 50% 243494 usec, 90% 496352 usec, 99% 588986 usec, 99.9% 604199 usec, 99.99% 604745 usec, max 604954 usec, measured over 1179648 events =====
```

**Comment by steveblackburn** [comment]
TL;DR this is not a bug. 71959c3 generalizes metered latency to provide a continuum from simple to metered, addressing the issue, I hope.

The best way to think of this is that metered latency moves all event start times so that they are evenly spaced on the benchmark timeline.  It's helpful to think of this as a *smoothing function* over the start times.  Then, we can consider what happens if we change the smoothing window.  If the window is as big as the whole execution, then the event starts are perfectly evenly spaced.  On the other hand if the window is arbitrarily small, then the event starts are unchanged.  These two extremes correspond to *simple* and *metered*.  What 71959c3 does is report results for different window sizes, revealing the effect of this smoothing function.  The result (below) makes it plain that the window size allows a transition between simple and metered: you can see the 50%-ile latency grow 547us -> 563us -> 770us -> 12589us -> 524272us as the smoothing window goes up by orders of magnitude.

The reason why the result grows to such a large extent in this case is because the underlying distribution of event start times is uneven.  I chose an example below that magnifies the effect (it is the first iteration).  To understand the behavior, consider an idealized situation where events are processed increasingly quickly. For example imagine the average event time shifts from 2ms to 500us over the duration of the benchmark.  If that were the case, the average time would be around 1ms.  If we were to evaluate the metered latency for such a case, for much of the first half of the benchmark, every event would be unable to complete on time and the imaginary event queue would grow and grow until eventually the events were being processed fast enough to drain the queue.  This would lead to enormous event latencies for events near the middle of the execution.

It's clear that simple latency is unrealistic because it does not model a queue, so the only events affected by a pause are those that are actually interrupted by the pause.  On the other hand it's clear that fully metered latency is unrealistic because it assumes an even distribution of events.  By generalizing the model to provide different smoothing functions, I hope users will form a clearer intuition into what the data means.  I can imagine that a smoothing function of something like 100ms might provide a reasonable middle ground between the simple and fully metered.

```
===== DaCapo evaluation-git-71959c39 cassandra PASSED in 10606 msec =====
===== DaCapo processed 200000 requests in 7490 msec, 26702 requests per second =====
===== DaCapo simple tail latency: 50% 547 usec, 90% 965 usec, 99% 1690 usec, 99.9% 4568 usec, 99.99% 29075 usec, max 51451 usec, measured over 200000 events =====
===== DaCapo 10ms metered tail latency: 50% 563 usec, 90% 1005 usec, 99% 2449 usec, 99.9% 27938 usec, 99.99% 60741 usec, max 67262 usec, measured over 200000 events =====
===== DaCapo 100ms metered tail latency: 50% 770 usec, 90% 3839 usec, 99% 19371 usec, 99.9% 96070 usec, 99.99% 100890 usec, max 107524 usec, measured over 200000 events =====
===== DaCapo 1000ms metered tail latency: 50% 12589 usec, 90% 102811 usec, 99% 183316 usec, 99.9% 185614 usec, 99.99% 186046 usec, max 186489 usec, measured over 200000 events =====
===== DaCapo 10000ms metered tail latency: 50% 524272 usec, 90% 770329 usec, 99% 779701 usec, 99.9% 780269 usec, 99.99% 780605 usec, max 780958 usec, measured over 200000 events =====

```

**Comment by caizixian** [comment]
Should we include one window that is`>= limitMS`, so we have at least one case that is the same as the old metered latency?

https://github.com/dacapobench/dacapobench/commit/131bdd228a52e2ff30010663a99070e4e5870538#diff-ae45bcd565bc49ca77b9a81285ec1e135d63f43c1e2f7f112fb3e8f34ab17753R237

**Comment by steveblackburn** [comment]
Yes, but the line above does exactly that, right?

```
limitMS = elapsedMS * 10;
```

(It multiplies the measured time by 10 and uses that as the sentinel, for the very reason you suggest.)

**Comment by caizixian** [comment]
> Yes, but the line above does exactly that, right?
> 
> ```
> limitMS = elapsedMS * 10;
> ```
> 
> (It multiplies the measured time by 10 and uses that as the sentinel, for the very reason you suggest.)

Of course. I missed that.

**Comment by steveblackburn** [comment]
Status: Although this was not a bug, it led to a a number of improvements, generalizations, and clarifications w.r.t. metered latency.

We now report simple, 100ms smoothed and fully smoothed by default.   We also have the `--latency-metered` option which will dump a csv with latencies at various smoothing windows from 100us up to the length of the benchmark execution.  This allows a detailed analysis of the systems' latency responsiveness.
